File: src/shadow.zig
// shadow: /etc/shadow line parser + hex decoder.
//
// Pure, allocation-free, no externs. The kernel's sys_authenticate
// (src/sys.zig) reads /etc/shadow into a stack buffer and walks it line by
// line with these helpers; the host tests below pin the format so a
// consumer never drifts from the build-time generator (tools/gen_shadow.zig).
//
// Line format: `user:iterations:salt_hex:hash_hex`
//   * user        — login name (raw bytes, compared verbatim)
//   * iterations  — PBKDF2-HMAC-SHA256 round count, decimal
//   * salt_hex    — salt bytes, hex (even length, lower/upper)
//   * hash_hex    — derived key bytes, hex (even length)
//
// salt/hash stay hex in the Entry; the caller hexDecode()s them right next
// to the PBKDF2 call, so this module owns no buffers. There is deliberately
// NO uid field — uid/gid/shell live in /etc/passwd (parsed in userland by
// /bin/login); /etc/shadow holds only the verifier, mirroring real Unix.

pub const Entry = struct {
    user: []const u8,
    iterations: u32,
    salt_hex: []const u8,
    hash_hex: []const u8,
};

// Split one shadow line (no trailing newline — the caller slices on '\n')
// into its four fields. Returns null on a missing or empty field, a 5th
// `:`-delimited field, or a non-decimal / zero / overflowing iteration
// count.
pub fn parseLine(line: []const u8) ?Entry {
    const c1 = indexOf(line, ':') orelse return null;
    const user = line[0..c1];
    const rest1 = line[c1 + 1 ..];

    const c2 = indexOf(rest1, ':') orelse return null;
    const iters_s = rest1[0..c2];
    const rest2 = rest1[c2 + 1 ..];

    const c3 = indexOf(rest2, ':') orelse return null;
    const salt_hex = rest2[0..c3];
    const hash_hex = rest2[c3 + 1 ..];

    // A 5th field (another ':') is malformed.
    if (indexOf(hash_hex, ':') != null) return null;
    if (user.len == 0 or iters_s.len == 0 or salt_hex.len == 0 or hash_hex.len == 0) return null;

    const iterations = parseDecimalU32(iters_s) orelse return null;
    if (iterations == 0) return null;

    return .{
        .user = user,
        .iterations = iterations,
        .salt_hex = salt_hex,
        .hash_hex = hash_hex,
    };
}

// Decode `in` (hex, even length) into `out`. Returns the byte count, or
// null on odd length, a non-hex digit, or `out` too small.
pub fn hexDecode(in: []const u8, out: []u8) ?usize {
    if (in.len % 2 != 0) return null;
    const n = in.len / 2;
    if (out.len < n) return null;
    var i: usize = 0;
    while (i < n) : (i += 1) {
        const hi = hexNibble(in[2 * i]) orelse return null;
        const lo = hexNibble(in[2 * i + 1]) orelse return null;
        out[i] = (hi << 4) | lo;
    }
    return n;
}

// Encode `in` bytes as lowercase hex into `out`. Returns the character
// count (2 × in.len), or null when `out` is too small. Inverse of
// hexDecode; sys_passwd uses it to serialize the fresh salt + derived
// key back into a shadow line.
pub fn hexEncode(in: []const u8, out: []u8) ?usize {
    if (out.len < in.len * 2) return null;
    const digits = "0123456789abcdef";
    for (in, 0..) |b, i| {
        out[2 * i] = digits[b >> 4];
        out[2 * i + 1] = digits[b & 0xF];
    }
    return in.len * 2;
}

// Byte span (start inclusive, end exclusive, newline excluded) of the
// shadow line whose user field equals `user`. Lines that fail parseLine
// are skipped, mirroring the lookup loop in sys_authenticate. Returns
// null when no line matches.
pub const LineSpan = struct { start: usize, end: usize };

pub fn findUserLine(content: []const u8, user: []const u8) ?LineSpan {
    var line_start: usize = 0;
    var i: usize = 0;
    while (i <= content.len) : (i += 1) {
        if (i != content.len and content[i] != '\n') continue;
        const line = content[line_start..i];
        const span_start = line_start;
        line_start = i + 1;
        if (line.len == 0) continue;
        const e = parseLine(line) orelse continue;
        if (bytesEqual(e.user, user)) return .{ .start = span_start, .end = i };
    }
    return null;
}

// Rewrite `user`'s shadow line in place with a fresh salt + hash, keeping
// the iteration count. The same-length invariant is what makes the
// follow-up FAT32 write splice-safe: the iteration count is reused
// verbatim and salt/hash arrive as fixed-width hex, so the new line is
// byte-for-byte the same length as the old one and the file size never
// changes. Returns false when the user is absent, the old line does not
// parse, or the lengths diverge (e.g. a hand-edited shadow with a
// different salt width — refuse rather than corrupt).
pub fn rewriteLineInPlace(
    content: []u8,
    user: []const u8,
    new_salt_hex: []const u8,
    new_hash_hex: []const u8,
) bool {
    const span = findUserLine(content, user) orelse return false;
    const old = parseLine(content[span.start..span.end]) orelse return false;

    const new_len = user.len + 1 + decimalLen(old.iterations) + 1 + new_salt_hex.len + 1 + new_hash_hex.len;
    if (new_len != span.end - span.start) return false;

    var w: usize = span.start;
    for (user) |c| {
        content[w] = c;
        w += 1;
    }
    content[w] = ':';
    w += 1;
    w += writeDecimal(content[w..], old.iterations);
    content[w] = ':';
    w += 1;
    for (new_salt_hex) |c| {
        content[w] = c;
        w += 1;
    }
    content[w] = ':';
    w += 1;
    for (new_hash_hex) |c| {
        content[w] = c;
        w += 1;
    }
    return w == span.end;
}

fn bytesEqual(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    for (a, b) |x, y| {
        if (x != y) return false;
    }
    return true;
}

// Digit count of `v` in decimal (v == 0 -> 1).
fn decimalLen(v: u32) usize {
    var n: usize = 1;
    var x = v / 10;
    while (x != 0) : (x /= 10) n += 1;
    return n;
}

// Write `v` in decimal at out[0..]; returns the digit count. The caller
// guarantees capacity (rewriteLineInPlace checked the total length).
fn writeDecimal(out: []u8, v: u32) usize {
    const n = decimalLen(v);
    var x = v;
    var i = n;
    while (i > 0) {
        i -= 1;
        out[i] = '0' + @as(u8, @intCast(x % 10));
        x /= 10;
    }
    return n;
}

fn indexOf(haystack: []const u8, needle: u8) ?usize {
    for (haystack, 0..) |c, i| {
        if (c == needle) return i;
    }
    return null;
}

// Decimal u32 parse, exact (no sign, no whitespace). A u64 accumulator
// catches overflow past u32 without depending on std.math.
fn parseDecimalU32(s: []const u8) ?u32 {
    var v: u64 = 0;
    for (s) |c| {
        if (c < '0' or c > '9') return null;
        v = v * 10 + (c - '0');
        if (v > 0xFFFF_FFFF) return null;
    }
    return @intCast(v);
}

fn hexNibble(c: u8) ?u8 {
    return switch (c) {
        '0'...'9' => c - '0',
        'a'...'f' => c - 'a' + 10,
        'A'...'F' => c - 'A' + 10,
        else => null,
    };
}

// ---- Host tests ----
const std = @import("std");

test "parseLine: well-formed line" {
    const e = parseLine("flash:4096:0011aabb:deadbeef").?;
    try std.testing.expectEqualStrings("flash", e.user);
    try std.testing.expectEqual(@as(u32, 4096), e.iterations);
    try std.testing.expectEqualStrings("0011aabb", e.salt_hex);
    try std.testing.expectEqualStrings("deadbeef", e.hash_hex);
}

test "parseLine: rejects missing fields" {
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash:4096:0011aabb"));
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash:4096"));
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash"));
    try std.testing.expectEqual(@as(?Entry, null), parseLine(""));
}

test "parseLine: rejects a 5th field" {
    try std.testing.expectEqual(@as(?Entry, null), parseLine("a:1:bb:cc:extra"));
}

test "parseLine: rejects empty user / non-decimal / zero iters" {
    try std.testing.expectEqual(@as(?Entry, null), parseLine(":4096:bb:cc"));
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash:40x6:bb:cc"));
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash:0:bb:cc"));
}

test "parseLine: rejects iteration overflow past u32" {
    try std.testing.expectEqual(@as(?Entry, null), parseLine("flash:99999999999:bb:cc"));
}

test "hexDecode: round-trips bytes" {
    var out: [4]u8 = undefined;
    const n = hexDecode("0011aabb", &out).?;
    try std.testing.expectEqual(@as(usize, 4), n);
    try std.testing.expectEqualSlices(u8, &[_]u8{ 0x00, 0x11, 0xAA, 0xBB }, out[0..n]);
}

test "hexDecode: accepts uppercase" {
    var out: [2]u8 = undefined;
    const n = hexDecode("DEAD", &out).?;
    try std.testing.expectEqualSlices(u8, &[_]u8{ 0xDE, 0xAD }, out[0..n]);
}

test "hexDecode: rejects odd length / bad digit / small out" {
    var out: [4]u8 = undefined;
    try std.testing.expectEqual(@as(?usize, null), hexDecode("abc", &out));
    try std.testing.expectEqual(@as(?usize, null), hexDecode("zz", &out));
    var small: [1]u8 = undefined;
    try std.testing.expectEqual(@as(?usize, null), hexDecode("aabb", &small));
}

test "hexEncode: lowercase round-trip with hexDecode" {
    const bytes = [_]u8{ 0x00, 0x11, 0xAA, 0xBB, 0xDE, 0xAD };
    var hex: [12]u8 = undefined;
    const n = hexEncode(&bytes, &hex).?;
    try std.testing.expectEqual(@as(usize, 12), n);
    try std.testing.expectEqualStrings("0011aabbdead", hex[0..n]);

    var back: [6]u8 = undefined;
    const m = hexDecode(hex[0..n], &back).?;
    try std.testing.expectEqualSlices(u8, &bytes, back[0..m]);
}

test "hexEncode: rejects an undersized output buffer" {
    const bytes = [_]u8{ 0x01, 0x02 };
    var small: [3]u8 = undefined;
    try std.testing.expectEqual(@as(?usize, null), hexEncode(&bytes, &small));
}

// Two-line fixture mirroring the gen_shadow output shape: 16-byte salts
// (32 hex chars) and 32-byte derived keys (64 hex chars).
const REWRITE_FIXTURE =
    "root:4096:" ++ ("aa" ** 16) ++ ":" ++ ("bb" ** 32) ++ "\n" ++
    "flash:4096:" ++ ("cc" ** 16) ++ ":" ++ ("dd" ** 32) ++ "\n";

test "findUserLine: locates first, last, and absent users" {
    const root_span = findUserLine(REWRITE_FIXTURE, "root").?;
    try std.testing.expectEqual(@as(usize, 0), root_span.start);
    const root_line = REWRITE_FIXTURE[root_span.start..root_span.end];
    try std.testing.expectEqualStrings("root", parseLine(root_line).?.user);

    const flash_span = findUserLine(REWRITE_FIXTURE, "flash").?;
    const flash_line = REWRITE_FIXTURE[flash_span.start..flash_span.end];
    try std.testing.expectEqualStrings("flash", parseLine(flash_line).?.user);
    // The span excludes the trailing newline.
    try std.testing.expectEqual(@as(u8, '\n'), REWRITE_FIXTURE[flash_span.end]);

    try std.testing.expectEqual(@as(?LineSpan, null), findUserLine(REWRITE_FIXTURE, "anton"));
    // A prefix of an existing user must not match.
    try std.testing.expectEqual(@as(?LineSpan, null), findUserLine(REWRITE_FIXTURE, "fla"));
}

test "findUserLine: works without a trailing newline on the last line" {
    const fixture = "root:4096:" ++ ("aa" ** 16) ++ ":" ++ ("bb" ** 32);
    const span = findUserLine(fixture, "root").?;
    try std.testing.expectEqual(fixture.len, span.end);
}

test "rewriteLineInPlace: same-length rewrite keeps neighbours and size intact" {
    var buf: [REWRITE_FIXTURE.len]u8 = undefined;
    @memcpy(&buf, REWRITE_FIXTURE);

    const new_salt = "0123456789abcdef0123456789abcdef"; // 32 hex chars
    const new_hash = "f0" ** 32; // 64 hex chars
    try std.testing.expect(rewriteLineInPlace(&buf, "flash", new_salt, new_hash));

    // The flash line carries the new salt + hash and still parses.
    const span = findUserLine(&buf, "flash").?;
    const e = parseLine(buf[span.start..span.end]).?;
    try std.testing.expectEqual(@as(u32, 4096), e.iterations);
    try std.testing.expectEqualStrings(new_salt, e.salt_hex);
    try std.testing.expectEqualStrings(new_hash, e.hash_hex);

    // The root line is byte-identical (no bleed across the rewrite).
    const root_span = findUserLine(&buf, "root").?;
    try std.testing.expectEqualStrings(
        REWRITE_FIXTURE[root_span.start..root_span.end],
        buf[root_span.start..root_span.end],
    );
    // Total content length is unchanged by construction (in-place).
}

test "rewriteLineInPlace: round-trips through PBKDF2-style fresh values twice" {
    // Two consecutive rewrites (the [TEST] passwd change + restore shape)
    // keep the file stable: same length, both parseable, order preserved.
    var buf: [REWRITE_FIXTURE.len]u8 = undefined;
    @memcpy(&buf, REWRITE_FIXTURE);

    try std.testing.expect(rewriteLineInPlace(&buf, "flash", "11" ** 16, "22" ** 32));
    try std.testing.expect(rewriteLineInPlace(&buf, "flash", "cc" ** 16, "dd" ** 32));
    try std.testing.expectEqualStrings(REWRITE_FIXTURE, &buf);
}

test "rewriteLineInPlace: refuses absent user and diverging lengths" {
    var buf: [REWRITE_FIXTURE.len]u8 = undefined;
    @memcpy(&buf, REWRITE_FIXTURE);

    // Absent user.
    try std.testing.expect(!rewriteLineInPlace(&buf, "anton", "aa" ** 16, "bb" ** 32));
    // Shorter salt (8 bytes hex = 16 chars) would shrink the line — refused.
    try std.testing.expect(!rewriteLineInPlace(&buf, "flash", "aa" ** 8, "bb" ** 32));
    // Longer hash would grow the line — refused.
    try std.testing.expect(!rewriteLineInPlace(&buf, "flash", "aa" ** 16, "bb" ** 33));
    // The refusals left the content untouched.
    try std.testing.expectEqualStrings(REWRITE_FIXTURE, &buf);
}

File: crates/kernel/src/security/shadow.rs
//! Allocation-free `/etc/shadow` parsing and in-place rewriting.

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Entry<'a> {
    pub user: &'a [u8],
    pub iterations: u32,
    pub salt_hex: &'a [u8],
    pub hash_hex: &'a [u8],
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct LineSpan {
    pub start: usize,
    pub end: usize,
}

pub fn parse_line(line: &[u8]) -> Option<Entry<'_>> {
    let c1 = index_of(line, b':')?;
    let user = &line[..c1];
    let rest1 = &line[c1 + 1..];
    let c2 = index_of(rest1, b':')?;
    let iterations_bytes = &rest1[..c2];
    let rest2 = &rest1[c2 + 1..];
    let c3 = index_of(rest2, b':')?;
    let salt_hex = &rest2[..c3];
    let hash_hex = &rest2[c3 + 1..];

    if hash_hex.contains(&b':')
        || user.is_empty()
        || iterations_bytes.is_empty()
        || salt_hex.is_empty()
        || hash_hex.is_empty()
    {
        return None;
    }
    let iterations = parse_decimal_u32(iterations_bytes)?;
    if iterations == 0 {
        return None;
    }
    Some(Entry {
        user,
        iterations,
        salt_hex,
        hash_hex,
    })
}

pub fn hex_decode(input: &[u8], out: &mut [u8]) -> Option<usize> {
    if !input.len().is_multiple_of(2) {
        return None;
    }
    let count = input.len() / 2;
    if out.len() < count {
        return None;
    }
    for i in 0..count {
        let hi = hex_nibble(input[2 * i])?;
        let lo = hex_nibble(input[2 * i + 1])?;
        out[i] = (hi << 4) | lo;
    }
    Some(count)
}

pub fn hex_encode(input: &[u8], out: &mut [u8]) -> Option<usize> {
    let count = input.len().checked_mul(2)?;
    if out.len() < count {
        return None;
    }
    const DIGITS: &[u8; 16] = b"0123456789abcdef";
    for (i, byte) in input.iter().copied().enumerate() {
        out[2 * i] = DIGITS[(byte >> 4) as usize];
        out[2 * i + 1] = DIGITS[(byte & 0x0f) as usize];
    }
    Some(count)
}

pub fn find_user_line(content: &[u8], user: &[u8]) -> Option<LineSpan> {
    let mut line_start = 0;
    let mut i = 0;
    while i <= content.len() {
        if i == content.len() || content[i] == b'\n' {
            let line = &content[line_start..i];
            let span_start = line_start;
            line_start = i + 1;
            if !line.is_empty() {
                if let Some(entry) = parse_line(line) {
                    if entry.user == user {
                        return Some(LineSpan {
                            start: span_start,
                            end: i,
                        });
                    }
                }
            }
        }
        i += 1;
    }
    None
}

pub fn rewrite_line_in_place(
    content: &mut [u8],
    user: &[u8],
    new_salt_hex: &[u8],
    new_hash_hex: &[u8],
) -> bool {
    let span = match find_user_line(content, user) {
        Some(span) => span,
        None => return false,
    };
    let iterations = match parse_line(&content[span.start..span.end]) {
        Some(entry) => entry.iterations,
        None => return false,
    };
    let new_len = match user
        .len()
        .checked_add(1)
        .and_then(|n| n.checked_add(decimal_len(iterations)))
        .and_then(|n| n.checked_add(1))
        .and_then(|n| n.checked_add(new_salt_hex.len()))
        .and_then(|n| n.checked_add(1))
        .and_then(|n| n.checked_add(new_hash_hex.len()))
    {
        Some(len) => len,
        None => return false,
    };
    if new_len != span.end - span.start {
        return false;
    }

    let mut write = span.start;
    content[write..write + user.len()].copy_from_slice(user);
    write += user.len();
    content[write] = b':';
    write += 1;
    write += write_decimal(&mut content[write..span.end], iterations);
    content[write] = b':';
    write += 1;
    content[write..write + new_salt_hex.len()].copy_from_slice(new_salt_hex);
    write += new_salt_hex.len();
    content[write] = b':';
    write += 1;
    content[write..write + new_hash_hex.len()].copy_from_slice(new_hash_hex);
    write += new_hash_hex.len();
    write == span.end
}

fn decimal_len(value: u32) -> usize {
    let mut count = 1;
    let mut rest = value / 10;
    while rest != 0 {
        count += 1;
        rest /= 10;
    }
    count
}

fn write_decimal(out: &mut [u8], value: u32) -> usize {
    let count = decimal_len(value);
    let mut rest = value;
    let mut i = count;
    while i > 0 {
        i -= 1;
        out[i] = b'0' + (rest % 10) as u8;
        rest /= 10;
    }
    count
}

fn index_of(bytes: &[u8], needle: u8) -> Option<usize> {
    bytes.iter().position(|byte| *byte == needle)
}

fn parse_decimal_u32(bytes: &[u8]) -> Option<u32> {
    let mut value = 0u64;
    for byte in bytes.iter().copied() {
        if !byte.is_ascii_digit() {
            return None;
        }
        value = value.checked_mul(10)?.checked_add(u64::from(byte - b'0'))?;
        if value > u64::from(u32::MAX) {
            return None;
        }
    }
    Some(value as u32)
}

fn hex_nibble(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        b'A'..=b'F' => Some(byte - b'A' + 10),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::{find_user_line, hex_decode, hex_encode, parse_line, rewrite_line_in_place};

    const REWRITE_FIXTURE: &[u8] = concat!(
        "root:4096:",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa:",
        "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb\n",
        "flash:4096:",
        "cccccccccccccccccccccccccccccccc:",
        "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd\n"
    )
    .as_bytes();

    #[test]
    fn parse_line_accepts_a_well_formed_line() {
        let entry = parse_line(b"flash:4096:0011aabb:deadbeef").unwrap();
        assert_eq!(entry.user, b"flash");
        assert_eq!(entry.iterations, 4096);
        assert_eq!(entry.salt_hex, b"0011aabb");
        assert_eq!(entry.hash_hex, b"deadbeef");
    }
    #[test]
    fn parse_line_rejects_missing_fields() {
        assert!(parse_line(b"flash:4096:0011aabb").is_none());
        assert!(parse_line(b"flash:4096").is_none());
        assert!(parse_line(b"flash").is_none());
        assert!(parse_line(b"").is_none());
    }
    #[test]
    fn parse_line_rejects_a_fifth_field() {
        assert!(parse_line(b"a:1:bb:cc:extra").is_none());
    }
    #[test]
    fn parse_line_rejects_empty_user_non_decimal_and_zero_iterations() {
        assert!(parse_line(b":4096:bb:cc").is_none());
        assert!(parse_line(b"flash:40x6:bb:cc").is_none());
        assert!(parse_line(b"flash:0:bb:cc").is_none());
    }
    #[test]
    fn parse_line_rejects_iteration_overflow_past_u32() {
        assert!(parse_line(b"flash:99999999999:bb:cc").is_none());
    }
    #[test]
    fn hex_decode_round_trips_bytes() {
        let mut out = [0u8; 4];
        let count = hex_decode(b"0011aabb", &mut out).unwrap();
        assert_eq!(count, 4);
        assert_eq!(&out[..count], &[0x00, 0x11, 0xaa, 0xbb]);
    }
    #[test]
    fn hex_decode_accepts_uppercase() {
        let mut out = [0u8; 2];
        let count = hex_decode(b"DEAD", &mut out).unwrap();
        assert_eq!(&out[..count], &[0xde, 0xad]);
    }
    #[test]
    fn hex_decode_rejects_odd_length_bad_digit_and_small_output() {
        let mut out = [0u8; 4];
        assert!(hex_decode(b"abc", &mut out).is_none());
        assert!(hex_decode(b"zz", &mut out).is_none());
        assert!(hex_decode(b"aabb", &mut out[..1]).is_none());
    }
    #[test]
    fn hex_encode_lowercase_round_trips_with_decode() {
        let bytes = [0x00, 0x11, 0xaa, 0xbb, 0xde, 0xad];
        let mut hex = [0u8; 12];
        let count = hex_encode(&bytes, &mut hex).unwrap();
        assert_eq!(&hex[..count], b"0011aabbdead");
        let mut back = [0u8; 6];
        let decoded = hex_decode(&hex[..count], &mut back).unwrap();
        assert_eq!(&back[..decoded], &bytes);
    }
    #[test]
    fn hex_encode_rejects_an_undersized_output_buffer() {
        let mut small = [0u8; 3];
        assert!(hex_encode(&[1, 2], &mut small).is_none());
    }
    #[test]
    fn find_user_line_locates_first_last_and_absent_users() {
        let root = find_user_line(REWRITE_FIXTURE, b"root").unwrap();
        assert_eq!(root.start, 0);
        assert_eq!(
            parse_line(&REWRITE_FIXTURE[root.start..root.end])
                .unwrap()
                .user,
            b"root"
        );
        let flash = find_user_line(REWRITE_FIXTURE, b"flash").unwrap();
        assert_eq!(
            parse_line(&REWRITE_FIXTURE[flash.start..flash.end])
                .unwrap()
                .user,
            b"flash"
        );
        assert_eq!(REWRITE_FIXTURE[flash.end], b'\n');
        assert!(find_user_line(REWRITE_FIXTURE, b"anton").is_none());
        assert!(find_user_line(REWRITE_FIXTURE, b"fla").is_none());
    }
    #[test]
    fn find_user_line_works_without_a_trailing_newline() {
        let fixture = b"root:4096:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb";
        assert_eq!(find_user_line(fixture, b"root").unwrap().end, fixture.len());
    }
    #[test]
    fn same_length_rewrite_keeps_neighbours_and_size_intact() {
        let mut content = REWRITE_FIXTURE.to_vec();
        let root = find_user_line(&content, b"root").unwrap();
        let root_before = content[root.start..root.end].to_vec();
        let salt = b"0123456789abcdef0123456789abcdef";
        let hash = [b'f'; 64];
        assert!(rewrite_line_in_place(&mut content, b"flash", salt, &hash));
        let flash = find_user_line(&content, b"flash").unwrap();
        let entry = parse_line(&content[flash.start..flash.end]).unwrap();
        assert_eq!(entry.iterations, 4096);
        assert_eq!(entry.salt_hex, salt);
        assert_eq!(entry.hash_hex, &hash);
        assert_eq!(&content[root.start..root.end], root_before);
        assert_eq!(content.len(), REWRITE_FIXTURE.len());
    }
    #[test]
    fn rewrite_round_trips_through_fresh_values_twice() {
        let mut content = REWRITE_FIXTURE.to_vec();
        assert!(rewrite_line_in_place(
            &mut content,
            b"flash",
            &[b'1'; 32],
            &[b'2'; 64]
        ));
        assert!(rewrite_line_in_place(
            &mut content,
            b"flash",
            &[b'c'; 32],
            &[b'd'; 64]
        ));
        assert_eq!(content, REWRITE_FIXTURE);
    }
    #[test]
    fn rewrite_refuses_absent_user_and_diverging_lengths() {
        let mut content = REWRITE_FIXTURE.to_vec();
        assert!(!rewrite_line_in_place(
            &mut content,
            b"anton",
            &[b'a'; 32],
            &[b'b'; 64]
        ));
        assert!(!rewrite_line_in_place(
            &mut content,
            b"flash",
            &[b'a'; 16],
            &[b'b'; 64]
        ));
        assert!(!rewrite_line_in_place(
            &mut content,
            b"flash",
            &[b'a'; 32],
            &[b'b'; 66]
        ));
        assert_eq!(content, REWRITE_FIXTURE);
    }
}
